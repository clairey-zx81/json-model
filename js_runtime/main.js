import { parseArgs } from 'node:util'
import fs from 'node:fs'

// TODO escaping?
const json_path = (path) => '.' + path.join('.')

function report(rep)
{
    if (rep === null || rep.length === 0) return ''

    let reps = []
    for (const [message, path] of rep)
        reps.push(`${json_path(path)}: ${message}`)

    return ' (' + reps.join('; ') + ')'
}

export default function main(checker)
{
    const options = {
      'verbose': { type: 'boolean', short: 'v' },
      'times': { type: 'string', short: 'T' },
      'report': { type: 'boolean', short: 'r' },
    }

    const args = parseArgs({options, allowPositionals: true})

    // args.hasOwnProperty('verbose')
    // args.verbose === undefined
    let times = args.values.times !== undefined ? Number(args.values.times) : 1

    if (isNaN(times))
    {
        console.log(`Unexpected --times: ${args.values.times}`)
        process.exit(1);
    }

    // overhead estimation
    let empty = 0.0
    if (times > 1)
    {
        let n = times
        while (n--)
        {
            const start = performance.now()
            empty += 1000.0 * (performance.now() - start)
        }
        empty /= times
    }

    for (const fname of args.positionals)
    { 
        try {
            fs.readFile(fname, (err, data) => {
                if (err) throw err
                const value = JSON.parse(data)

                // show performance
                // NOTE: jit, calls may take a lot of time at the beginning
                // NOTE: gc, may be irregular…
                if (times > 1)
                {
                    let sum = 0.0
                    let sum2 = 0.0
                    let n = times
                    while (n--) {
                        const start = performance.now()
                        let rep = args.values.report ? [] : null
                        const valid = checker(value, '', rep)
                        rep && (rep.length = 0)
                        let delay = (1000.0 * (performance.now() - start)) - empty  // µs
                        sum += delay
                        sum2 += delay * delay
                    }
                    const avg = sum / times
                    const stdev = Math.sqrt(sum2 / times - avg * avg)
                    console.log(`${fname}: ${args.values.report ? 'rep' : 'nop'}`,
                                `${avg.toFixed(3)} ± ${stdev.toFixed(3)} µs (${empty.toFixed(3)})`)
                }

                // do it anyway for the result
                let rep = args.values.report ? [] : null
                const valid = checker(value, '', rep)
                console.log(`${fname}: ${valid ? 'PASS' : 'FAIL'}${report(rep)}`)
            })
        }
        catch (e) {
            console.log(`${fname}: ERROR (${e})`)
        }
    }
}
