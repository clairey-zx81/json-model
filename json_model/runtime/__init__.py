from .types import (
    Jsonable,
    Path,
    Report,
    PropMap,
    CheckFun,
    RegexFun,
    EntryCheckFun,
)

from .support import (
    # TODO remove _
    UNDEFINED,
    _tname,
    _path,
    _rep,
    Const,
    ConstSet,
    ConstMap,
    is_valid_date,
    is_valid_time,
    is_valid_datetime,
    is_valid_regex,
    is_valid_exreg,
    is_valid_url,
    is_valid_email,
    is_valid_uuid,
    is_valid_json,
    is_unique_array,
    check_constraint,
    value_len,
    main,
)
