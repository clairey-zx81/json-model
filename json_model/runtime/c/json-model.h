#ifndef JSON_MODEL_H_INCLUDED
#define JSON_MODEL_H_INCLUDED

#define _GNU_SOURCE  // for qsort_r?
#include <stdbool.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

#include <jansson.h>

/*
 * build generated API names
 */
#ifndef CHECK_FUNCTION_NAME
#warning defining CHECK_FUNCTION_NAME macro
#define CHECK_FUNCTION_NAME check_model
#endif  // !CHECK_FUNCTION_NAME

// branch hints
#define likely(x)       __builtin_expect(!!(x), 1)
#define unlikely(x)     __builtin_expect(!!(x), 0)

// inlining hints
#if defined(NO_INLINE_HINT)
#define INLINE
#else
#if defined(__clang__)
#define INLINE __attribute__((always_inline))
#elif defined(__GNUC__)
#define INLINE inline
// #elif defined(_MSC_VER)
#else
#define INLINE
#endif  // clang/gcc/...
#endif  // NO_INLINE_HINT

#define concat(prefix, suffix) prefix ## suffix
#define newname(prefix, suffix) concat(prefix, suffix)

#define CHECK_map_size newname(CHECK_FUNCTION_NAME, _map_size)
#define CHECK_map_tab newname(CHECK_FUNCTION_NAME, _map_tab)
#define CHECK_fun newname(CHECK_FUNCTION_NAME, _map)
#define CHECK_init newname(CHECK_FUNCTION_NAME, _init)
#define CHECK CHECK_FUNCTION_NAME
#define CHECK_free newname(CHECK_FUNCTION_NAME, _free)

// set on init
extern char * jm_version_string;

#define array_length(arr) (sizeof(arr) / sizeof((arr)[0]))
typedef int (*jm_cmp_fun_t)(const void *, const void *);

/*
 * JSON API extensions
 */

// keep object prop/vals as an array for sorting and fast get.
typedef struct {
    const char *prop;
    const json_t *val;
} jm_propval_t;

extern int jm_propval_cmp(const jm_propval_t *, const jm_propval_t *);

extern bool jm_json_is_scalar(const json_t *);
extern int jm_json_cmp(const json_t *, const json_t *);
extern int jm_json_array_cmp(const json_t *, const json_t *);
extern int jm_json_object_cmp(const json_t *, const json_t *);
extern bool jm_json_array_unique(const json_t *);

/*
 * reporting
 */
// linked list of report entries
typedef struct jm_report_entry {
  const char *message;
  char *path;
  struct jm_report_entry *prev;
} jm_report_entry_t;

typedef struct {
  jm_report_entry_t *entry;
} jm_report_t;

// path segment, name == NULL means use index
typedef struct jm_path {
    const char *name;
    size_t index;
    struct jm_path *prev;  // link to previous (in stack) segment
    struct jm_path *next;  // used for backtacking
} jm_path_t;

extern void jm_report_add_entry(jm_report_t* rep, const char *msg, jm_path_t *path);
extern void jm_report_free_entries(jm_report_t *rep);

typedef bool (*jm_check_fun_t)(const json_t *, jm_path_t *, jm_report_t *);

extern bool jm_check_fun_string(jm_check_fun_t, const char *, jm_path_t *, jm_report_t *);
extern bool jm_array_is_unique(const json_t *, jm_path_t *, jm_report_t *);

/*
 * property mapping management
 */
typedef struct {
    const char *name;
    jm_check_fun_t val_check;
} jm_propmap_t;

extern void jm_sort_propmap(jm_propmap_t *, int);
extern jm_check_fun_t jm_search_propmap(const char *, const jm_propmap_t *, int);

/*
 * set of scalar constants
 */
typedef enum {
    cst_is_null,
    cst_is_bool,
    cst_is_integer,
    cst_is_float,
    cst_is_string
    // cst_is_array + const json_t * + free
    // cst_is_object + const json_t * + free
} jm_constant_tag;

typedef struct
{
    jm_constant_tag tag;
    union {
        bool b;
        int64_t i;
        double f;
        const char *s;
    } val;
} jm_constant_t;

extern bool jm_set_cst(jm_constant_t *, const json_t *);
extern void jm_dbg_cst(const jm_constant_t *, const char *);
extern void jm_sort_cst(jm_constant_t *, size_t);
extern bool jm_search_cst(const jm_constant_t *, const jm_constant_t *, size_t);

/*
 * constant mapping management for discriminant optimization
 */
typedef struct {
    jm_constant_t cst;
    jm_check_fun_t check_val;
} jm_constmap_t;

extern void jm_sort_constmap(jm_constmap_t *, size_t);
extern jm_check_fun_t jm_search_constmap(const jm_constant_t *, const jm_constmap_t *, size_t);

/*
 * Miscellaneous support functions and related data
 */
extern bool jm_is_valid_date_slow(const char *, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_date_fast(const char *, jm_path_t *, jm_report_t *);
extern bool (*jm_is_valid_date)(const char *, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_time(const char *, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_datetime(const char *, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_uuid(const char *, jm_path_t *, jm_report_t *);
extern size_t jm_str_len(const char *);
extern size_t jm_any_len(json_t *);
extern bool jm_is_valid_regex_slow(const char *, bool, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_regex_fast(const char *, bool, jm_path_t *, jm_report_t *);
extern bool (*jm_is_valid_regex)(const char *, bool, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_url(const char *, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_email(const char *, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_json(const char *, jm_path_t *, jm_report_t *);

typedef enum {
    op_eq,
    op_ne,
    op_le,
    op_lt,
    op_ge,
    op_gt
} jm_constraint_op_t;

extern bool
jm_check_constraint(const json_t *, jm_constraint_op_t, const jm_constant_t *,
                    jm_path_t *, jm_report_t *);

/*
 * Shared high-level entry point
 */
extern bool jm_generic_entry(const char *(*)(void), jm_check_fun_t (*)(const char *),
                             const json_t *, const char *, bool *, char **);

/*
 * Generated stuff
 */
// low-level interface
extern const size_t CHECK_map_size;
extern jm_propmap_t CHECK_map_tab[];
extern jm_check_fun_t CHECK_fun(const char *);
// high-level interface
extern const char *CHECK_init(void);
extern bool CHECK(const json_t *, const char *, bool *, char **);
extern void CHECK_free(void);

#endif  // JSON_MODEL_H_INCLUDED
