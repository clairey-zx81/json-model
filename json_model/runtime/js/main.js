import { parseArgs } from 'node:util'
import fs from 'node:fs/promises'

import { jm_set_rx } from 'json_model_runtime'
import { createRequire } from 'node:module';
const require = createRequire(import.meta.url);

// TODO escaping?
const json_path = (path) => '.' + path.join('.')

function reporting(rep)
{
    if (rep === null || rep.length === 0) return ''

    let reps = []
    for (const [message, path] of rep)
        reps.unshift(`${json_path(path)}: ${message}`)

    return ' (' + reps.join('; ') + ')'
}

// process one value
function processing(source, value, checker, name, expected, report, times, empty)
{
    // do it once for the result
    let rep = report ? [] : null
    const valid = checker(value, name, rep)
    const error = (expected === null || valid === expected) ? 0 : 1

    if (error)
        console.log(`${source}: ERROR unexpected ${valid ? 'PASS' : 'FAIL'}${reporting(rep)}`)
    else
        console.log(`${source}: ${valid ? 'PASS' : 'FAIL'}${reporting(rep)}`)

    // show performance
    // NOTE: jit, calls may take a lot of time at the beginning
    // NOTE: gc, may be irregular…
    if (times > 1)
    {
        let sum = 0.0
        let sum2 = 0.0
        let n = times
        while (n--) {
            let rep = report ? [] : null
            const start = performance.now()
            checker(value, '', rep)
            rep && (rep.length = 0)
            let delay = (1000.0 * (performance.now() - start)) - empty  // µs
            sum += delay
            sum2 += delay * delay
        }
        const avg = sum / times
        const stdev = Math.sqrt(sum2 / times - avg * avg)
        console.warn(`${source}: ${report ? 'rep' : 'nop'} ${valid ? 'PASS' : 'FAIL'}`,
                     `${avg.toFixed(3)} ± ${stdev.toFixed(3)} µs (${empty.toFixed(3)})`)
    }

    return error
}

// for https://github.com/sourcemeta-research/jsonschema-benchmark
function jsonschema_benchmark(values, checker, times)
{
    let errors = 0

    // cold run
    const cold_start = performance.now()
    for (const v of values)
        if (!checker(v, "", null))
            errors++
    const cold_delay = performance.now() - cold_start  // ms

    const WARMUP_MAX_TIME = 10.0  // seconds
    const WARMUP_ITERATIONS = 100  // unless too long
    const max_iterations = Math.ceil(1000.0 * WARMUP_MAX_TIME / cold_delay)

    // warm-up so as to trigger JIT
    const warmup_iterations = Math.min(WARMUP_ITERATIONS, max_iterations)
    console.error(`warmup iterations: ${warmup_iterations}`)
    for (let i = 0; i < warmup_iterations; i++)
        for (const v of values)
            checker(v, '', null)

    // warm run
    let sum1 = 0.0, sum2 = 0.0
    for (let i = 0; i < times; i++)
    {
        const start = performance.now()
        for (const v of values)
            checker(v, "", null)
        const delay = performance.now() - start  // ms
        sum1 += delay
        sum2 += delay * delay
    }

    // results
    const avg = sum1 / times
    const stdev = Math.sqrt(sum2 / times - avg * avg)
    console.error(`js validation: pass=${values.length - errors} fail=${errors}`,
                  `${(1000.0 * avg).toFixed(3)} ± ${(1000.0 * stdev).toFixed(3)} µs`)

    console.log((1000000.0 * cold_delay).toFixed(0) + ',' + (1000000.0 * avg).toFixed(0))
    process.exit(errors ? 1 : 0)
}

// evaluate measure overhead
function eval_empty(times)
{
    let empty = 0.0
    if (times >= 1)
    {
        let n = times
        while (n--)
        {
            const start = performance.now()
            empty += 1000.0 * (performance.now() - start)
        }
        empty /= times
    }
    return empty
}

export default async function main(checker_init, checker, checker_free)
{
    checker_init()

    const options = {
      // TODO --help --list
      'verbose': { type: 'boolean', short: 'v' },
      'time': { type: 'string', short: 'T' },
      'report': { type: 'boolean', short: 'r' },
      'no-report': { type: 'boolean' },
      'test': { type: 'boolean', short: 't' },
      'jsonl': { type: 'boolean', short: 'L' },
      're2': { type: 'boolean' },
      'regexp': { type: 'boolean' },
      'jsonschema-benchmark': { type: 'boolean', short: 'B' },
    }

    const args = parseArgs({options, allowPositionals: true})

    const benchmarking = args.values["jsonschema-benchmark"]
    if (benchmarking)
        args.values.jsonl = true
    if (!benchmarking && args.values.jsonl)
        args.values.test = true
    if (args.values.no_report)
        args.values.report = false

    let errors = 0
    const times = args.values.time !== undefined ? Number(args.values.time) : 1

    if (isNaN(times))
    {
        console.log(`Unexpected --time: ${args.values.time}`)
        process.exit(1);
    }

    // reset regular expression engine
    if (args.values.regexp)
        jm_set_rx(RegExp)
    if (args.values.re2)
        jm_set_rx(require('re2'))

    // overhead estimation… we need the jit to kick in
    const empty = Math.min(
        eval_empty(times), eval_empty(times), eval_empty(times),
        eval_empty(times), eval_empty(times), eval_empty(times)
    )

    for (const fname of args.positionals)
    {
        try {
            const data = await fs.readFile(fname, {encoding: 'UTF-8'})
            const value = (
                args.values.jsonl ?
                    args.values.test ?
                        data.split("\n").slice(0, -1).map(s => [null, JSON.parse(s)])
                      : data.split("\n").slice(0, -1).map(s => JSON.parse(s))
                  : JSON.parse(data)
            )

            if (benchmarking)
                jsonschema_benchmark(value, checker, times)

            if (args.values.test) {
                if (!Array.isArray(value)) {
                    errors++
                    console.log(`${fname}: ERROR expecting an array`)
                }

                let index = 0
                for (const item of value)
                {
                    const display = `${fname}[${args.values.jsonl ? index+1 : index}]`

                    if (typeof item === 'string' || item instanceof String)  // skip comments
                        continue
                    if (!Array.isArray(item) || (item.length < 2 || item.length > 3)) {
                        errors++
                        console.log(`${display}: ERROR expecting a 2-tuple or 3-tuple`)
                        continue
                    }

                    let expect, name, val
                    if (item.length == 3)
                        [expect, name, val] = item
                    else
                        [expect, val] = item, name = ''

                    if (typeof expect !== 'boolean' && ! expect instanceof Boolean) {
                        errors++
                        console.log(`${display}: ERROR expecting a boolean first item`)
                        continue
                    }
                    if (typeof name !== 'string' && ! name instanceof String) {
                        errors++
                        console.log(`${display}: ERROR expecting a string for name`)
                        continue
                    }

                    errors += processing(display, val, checker, name,
                                         expect, args.values.report, times, empty)

                    index += 1
                }
            }
            else
                errors += processing(fname, value, checker, '',
                                     null, args.values.report, times, empty)
        }
        catch (e) {
            errors++
            console.log(`${fname}: ERROR (${String(e).replace('\n', '\\n')})`)
        }
    }

    checker_free()
    process.exit(errors ? 4 : 0)
}
