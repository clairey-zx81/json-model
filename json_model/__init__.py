#
# convenient types, classes and functions to process models
#
from .mtypes import Jsonable
from .runtime.types import Report, EntryCheckFun
from .model import JsonModel
from .script import (
    jmc_script,
    model_from_json, model_from_url, model_from_str,
    model_checker, model_checker_from_url, model_checker_from_json,
)

__ALL__ = [
    Jsonable, Report, EntryCheckFun, JsonModel,
    jmc_script,
    model_from_json, model_from_url, model_from_str,
    model_checker, model_checker_from_url, model_checker_from_json,
]
