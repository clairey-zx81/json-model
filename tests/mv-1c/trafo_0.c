//
// Generated by JSON Model Compiler version 2.0b0
// see https://github.com/clairey-zx81/json-model
//

// regular expression engine
#include <stddef.h>
#include <cre2.h>

#include <json-model.h>
#define JSON_MODEL_VERSION "2.0b0"

static bool json_model_2(const json_t *val, jm_path_t *path, jm_report_t *rep);
static bool json_model_1(const json_t *val, jm_path_t *path, jm_report_t *rep);
jm_propmap_t check_model_map_tab[2];
const size_t check_model_map_size = 2;

// object .'$zero'
static bool _jm_obj_0(const json_t *val, jm_path_t *path, jm_report_t *rep)
{
    if (! json_is_object(val))
    {
        if (rep) jm_report_add_entry(rep, "not an object [.'$zero']", path);
        return false;
    }
    bool res;
    int64_t must_count = 0;
    const char *prop;
    json_t *pval;
    json_object_foreach((json_t *) val, prop, pval)
    {
        jm_path_t lpath_0 = (jm_path_t) { prop, 0, path, NULL };
        if (strcmp(prop, "za") == 0)
        {
            // handle must za property
            must_count += 1;
            // .'$zero'.za
            res = json_is_integer(pval) && json_integer_value(pval) >= 0;
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "not a 0 strict int [.'$zero'.za]", (path ? &lpath_0 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid mandatory prop value [.'$zero'.za]", (path ? &lpath_0 : NULL));
                return false;
            }
        }
        else if (strcmp(prop, "zb") == 0)
        {
            // handle must zb property
            must_count += 1;
            // .'$zero'.zb
            res = json_is_integer(pval) && json_integer_value(pval) >= 0;
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "not a 0 strict int [.'$zero'.zb]", (path ? &lpath_0 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid mandatory prop value [.'$zero'.zb]", (path ? &lpath_0 : NULL));
                return false;
            }
        }
        else
        {
            if (rep) jm_report_add_entry(rep, "unexpected prop [.'$zero']", (path ? &lpath_0 : NULL));
            return false;
        }
    }
    if (must_count != 2)
    {
        if (rep != NULL)
        {
            if (! (json_object_get(val, "za") != NULL))
            {
                if (rep) jm_report_add_entry(rep, "missing mandatory prop <za> [.'$zero']", path);
            }
            if (! (json_object_get(val, "zb") != NULL))
            {
                if (rep) jm_report_add_entry(rep, "missing mandatory prop <zb> [.'$zero']", path);
            }
        }
        return false;
    }
    return true;
}

// check $zero (.'$zero')
static bool json_model_2(const json_t *val, jm_path_t *path, jm_report_t *rep)
{
    bool res;
    // .'$zero'
    res = _jm_obj_0(val, path, rep);
    if (! res)
    {
        if (rep) jm_report_add_entry(rep, "unexpected object [.'$zero']", path);
    }
    return res;
}

// check $ (.)
static bool json_model_1(const json_t *val, jm_path_t *path, jm_report_t *rep)
{
    bool res;
    // .
    res = json_model_2(val, path, rep);
    if (! res)
    {
        if (rep) jm_report_add_entry(rep, "unexpected $zero [.]", path);
    }
    return res;
}

jm_check_fun_t check_model_map(const char *pname)
{
    return jm_search_propmap(pname, check_model_map_tab, 2);
}

static bool initialized = false;

const char *check_model_init(void)
{
    if (!initialized)
    {
        initialized = true;
        jm_version_string = JSON_MODEL_VERSION;
        check_model_map_tab[0] = (jm_propmap_t) { "", json_model_1 };
        check_model_map_tab[1] = (jm_propmap_t) { "zero", json_model_2 };
        jm_sort_propmap(check_model_map_tab, 2);
    }
    return NULL;
}

void check_model_free(void)
{
    if (initialized)
    {
        initialized = false;

        // cleanup code
    }
}


/*
 * # API
 *
 * ```c
 * bool error;
 * char *message;
 *
 * bool valid = check(json_value, "model-name", &error, &message);
 *
 * if (message)
 *    free(message), message = NULL;
 * ```
 *
 * - `error` is set of there was an error during initialization or if the model does not exists.
 * - `message` is provided if a non NULL pointer is passed, and must be freed by the caller.
 */
bool
check_model(const json_t *val, const char *name, bool *error, char **reasons)
{
    return jm_generic_entry(check_model_init, check_model_map, val, name, error, reasons);
}
