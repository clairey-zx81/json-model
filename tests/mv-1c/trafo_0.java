//
// Generated by JSON Model Compiler version 2
// see https://github.com/clairey-zx81/json-model
//

import json_model.*;
import java.util.Map;
import java.util.Set;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

@SuppressWarnings("unchecked")
public class trafo_0 extends ModelChecker
{
    static public final String VERSION = "2";

    public Map<String, Checker> trafo_0_map_pmap;

    // object .'$zero'
    public boolean _jm_obj_0(Object val, Path path, Report rep)
    {
        // check close must only props
        if (! json.isObject(val))
        {
            if (rep != null) rep.addEntry("not an object [.'$zero']", path);
            return false;
        }
        if (json.objectSize(val) != 2)
        {
            if (rep != null) rep.addEntry("bad property count [.'$zero']", path);
            return false;
        }
        Object pval;
        boolean res;
        if (! json.objectHasProp(val, "za"))
        {
            if (rep != null) rep.addEntry("missing mandatory prop <za> [.'$zero']", path);
            return false;
        }
        pval = json.objectValue(val, "za");
        // .'$zero'.za
        res = json.isInteger(pval) && json.asLong(pval) >= 0;
        if (! res)
        {
            if (rep != null) rep.addEntry("not a 0 strict int [.'$zero'.za]", path);
            if (rep != null) rep.addEntry("unexpected value for mandatory prop <za> [.'$zero']", path);
            return false;
        }
        if (! json.objectHasProp(val, "zb"))
        {
            if (rep != null) rep.addEntry("missing mandatory prop <zb> [.'$zero']", path);
            return false;
        }
        pval = json.objectValue(val, "zb");
        // .'$zero'.zb
        res = json.isInteger(pval) && json.asLong(pval) >= 0;
        if (! res)
        {
            if (rep != null) rep.addEntry("not a 0 strict int [.'$zero'.zb]", path);
            if (rep != null) rep.addEntry("unexpected value for mandatory prop <zb> [.'$zero']", path);
            return false;
        }
        return true;
    }

    // check $zero (.'$zero')
    public boolean json_model_2(Object val, Path path, Report rep)
    {
        boolean res;
        // .'$zero'
        res = _jm_obj_0(val, path, rep);
        if (! res)
        {
            if (rep != null) rep.addEntry("unexpected element [.'$zero']", path);
        }
        return res;
    }

    // check $ (.)
    public boolean json_model_1(Object val, Path path, Report rep)
    {
        boolean res;
        // .
        res = _jm_obj_0(val, path, rep);
        if (! res)
        {
            if (rep != null) rep.addEntry("unexpected $zero [.]", path);
        }
        return res;
    }


    public void init(JSON json)
    {
        if (!initialized)
        {
            try {
            trafo_0_map_pmap = new HashMap<String, Checker>();
            trafo_0_map_pmap.put("", new Checker() { public boolean call(Object o, Path p, Report r) { return _jm_obj_0(o, p, r);} });
            trafo_0_map_pmap.put("zero", new Checker() { public boolean call(Object o, Path p, Report r) { return _jm_obj_0(o, p, r);} });
                super.init(json);
            }
            catch (Exception e) {
                throw new Error("cannot initialized model checker: " + e);
            }
        }
    }

    public void free()
    {
        if (initialized)
        {
            super.free();
            trafo_0_map_pmap = null;
        }
    }

    public Checker get(String name)
    {
        return trafo_0_map_pmap.get(name);
    }

    public Set<String> models()
    {
        return trafo_0_map_pmap.keySet();
    }

    static public void main(String[] args) throws Exception
    {
        ModelChecker checker = new trafo_0();
        Main.main("trafo_0", checker, VERSION, args);
    }
}
