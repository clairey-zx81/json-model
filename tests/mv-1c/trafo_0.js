#! /bin/env node
// we may need require to load re2
import { createRequire } from 'node:module';
const require = createRequire(import.meta.url);
//
// Generated by JSON Model Compiler version 2
// see https://github.com/clairey-zx81/json-model
//
import * as runtime from "json_model_runtime"
const JSON_MODEL_VERSION = "2";

var check_model_map = new Map()

// object .'$zero'
function _jm_obj_0(val, path, rep)
{
    // check close must only props
    if (! (Object.prototype.toString.call(val) === '[object Object]'))
    {
        rep !== null && rep.push(["not an object [.'$zero']", path])
        return false;
    }
    if (Object.keys(val).length != 2)
    {
        rep !== null && rep.push(["bad property count [.'$zero']", path])
        return false;
    }
    let lpath;
    let pval;
    let res;
    if (! val.hasOwnProperty("za"))
    {
        rep !== null && rep.push(["missing mandatory prop <za> [.'$zero']", path])
        return false;
    }
    lpath = path ? path.concat(["za"]) : null;
    pval = val["za"];
    // .'$zero'.za
    res = ((typeof pval === 'number' || pval instanceof Number) && Number.isInteger(pval)) && pval >= 0;
    if (! res)
    {
        rep !== null && rep.push(["not a 0 strict int [.'$zero'.za]", (path ? lpath : null)])
        rep !== null && rep.push(["unexpected value for mandatory prop <za> [.'$zero']", (path ? lpath : null)])
        return false;
    }
    if (! val.hasOwnProperty("zb"))
    {
        rep !== null && rep.push(["missing mandatory prop <zb> [.'$zero']", path])
        return false;
    }
    lpath = path ? path.concat(["zb"]) : null;
    pval = val["zb"];
    // .'$zero'.zb
    res = ((typeof pval === 'number' || pval instanceof Number) && Number.isInteger(pval)) && pval >= 0;
    if (! res)
    {
        rep !== null && rep.push(["not a 0 strict int [.'$zero'.zb]", (path ? lpath : null)])
        rep !== null && rep.push(["unexpected value for mandatory prop <zb> [.'$zero']", (path ? lpath : null)])
        return false;
    }
    return true;
}

// check $zero (.'$zero')
function json_model_2(val, path, rep)
{
    let res;
    // .'$zero'
    res = _jm_obj_0(val, path, rep);
    if (! res)
    {
        rep !== null && rep.push(["unexpected element [.'$zero']", path])
    }
    return res;
}

// check $ (.)
function json_model_1(val, path, rep)
{
    let res;
    // .
    res = _jm_obj_0(val, path, rep);
    if (! res)
    {
        rep !== null && rep.push(["unexpected $zero [.]", path])
    }
    return res;
}


var initialized = false

// differed module initializations
export function check_model_init()
{
    if (! initialized)
    {
        initialized = true;
        runtime.jm_set_rx(RegExp)
        check_model_map.set("", _jm_obj_0)
        check_model_map.set("zero", _jm_obj_0)
    }
}

// differed module cleanup
export function check_model_free()
{
    if (initialized)
    {
        initialized = false;
    }
}

export function check_model(val, name, rep)
{
    let checker = check_model_map.get(name)
    if (checker === undefined)
        throw `no checker for "${name}"`
    let path = rep !== null ? [] : null

    return checker(val, path, rep)
}

// possibly run as main based on a guess
import main from "json_model_runtime/main.js"

if (import.meta.url.endsWith(process.argv[1]))
    main(check_model_init, check_model, check_model_free)
