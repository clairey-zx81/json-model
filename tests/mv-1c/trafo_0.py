#! /bin/env python
#
# Generated by JSON Model Compiler version 2
# see https://github.com/clairey-zx81/json-model
#
from typing import Callable
import re2 as re
from json_model.runtime import *
__version__ = "2"

def check_model_fun(name: str) -> CheckFun:
    """Return check function for JSON model name."""
    return check_model_map[name]

# entry point for generated checkers
def check_model(val: Jsonable, name: str = "", rep: Report = None) -> bool:
    """Check val validity agains JSON Model name."""
    if name not in check_model_map:
        raise Exception(f"unexpected model name: {name}")
    checker = check_model_map[name]
    return checker(val, [], rep)

check_model_map: PropMap

# check $zero (.'$zero')
def json_model_2(val: Jsonable, path: Path, rep: Report) -> bool:
    # .'$zero'
    # check close must only props
    if not isinstance(val, dict):
        rep is None or rep.append(("not an object [.'$zero']", path))
        return False
    if len(val) != 2:
        rep is None or rep.append(("bad property count [.'$zero']", path))
        return False
    lpath: Path
    pval: Jsonable
    res: bool
    if not ((pval := val.get("za", UNDEFINED)) != UNDEFINED):
        rep is None or rep.append(("missing mandatory prop <za> [.'$zero']", path))
        return False
    lpath = (path + [ "za" ]) if path is not None else None
    # .'$zero'.za
    res = isinstance(pval, int) and not isinstance(pval, bool) and pval >= 0
    if not res:
        rep is None or rep.append(("not a 0 strict int [.'$zero'.za]", lpath if path is not None else None))
        rep is None or rep.append(("unexpected value for mandatory prop <za> [.'$zero']", lpath if path is not None else None))
        return False
    if not ((pval := val.get("zb", UNDEFINED)) != UNDEFINED):
        rep is None or rep.append(("missing mandatory prop <zb> [.'$zero']", path))
        return False
    lpath = (path + [ "zb" ]) if path is not None else None
    # .'$zero'.zb
    res = isinstance(pval, int) and not isinstance(pval, bool) and pval >= 0
    if not res:
        rep is None or rep.append(("not a 0 strict int [.'$zero'.zb]", lpath if path is not None else None))
        rep is None or rep.append(("unexpected value for mandatory prop <zb> [.'$zero']", lpath if path is not None else None))
        return False
    return True

# check $ (.)
def json_model_1(val: Jsonable, path: Path, rep: Report) -> bool:
    res: bool
    # .
    res = json_model_2(val, path, rep)
    if not res:
        rep is None or rep.append(("unexpected $zero [.]", path))
    return res


# initialization guard
initialized: bool = False

# differed module initializations
def check_model_init():
    global initialized
    if not initialized:
        initialized = True
        global check_model_map
        check_model_map = {
            "": json_model_2,
            "zero": json_model_2,
        }

# differed module cleanup
def check_model_free():
    global initialized
    if initialized:
        initialized = False

if __name__ == "__main__":
    check_model_init()
    main(check_model_fun, check_model_map, __version__)
    check_model_free()
