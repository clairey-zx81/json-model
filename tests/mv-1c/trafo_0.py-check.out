trafo_0.values.json[0]: PASS
trafo_0.values.json[1]: FAIL [("missing must prop <za> [.'$zero']", []), ("missing must prop <zb> [.'$zero']", []), ("not an expected object at [.'$zero']", []), ('unexpected $zero []', [])]
trafo_0.values.json[2]: FAIL [("missing must prop <zb> [.'$zero']", []), ("not an expected object at [.'$zero']", []), ('unexpected $zero []', [])]
trafo_0.values.json[3]: FAIL [("missing must prop <za> [.'$zero']", []), ("not an expected object at [.'$zero']", []), ('unexpected $zero []', [])]
trafo_0.values.json[4]: FAIL [("no other prop expected [.'$zero']", ['un']), ("not an expected object at [.'$zero']", []), ('unexpected $zero []', [])]
trafo_0.values.json[5]: FAIL [("not an object [.'$zero']", []), ("not an expected object at [.'$zero']", []), ('unexpected $zero []', [])]
trafo_0.values.json[6]: FAIL [("not an object [.'$zero']", []), ("not an expected object at [.'$zero']", []), ('unexpected $zero []', [])]
trafo_0.values.json[7]: FAIL [("not an object [.'$zero']", []), ("not an expected object at [.'$zero']", []), ('unexpected $zero []', [])]
trafo_0.values.json[8]: FAIL [("not an object [.'$zero']", []), ("not an expected object at [.'$zero']", []), ('unexpected $zero []', [])]
trafo_0.values.json[9]: FAIL [("not an object [.'$zero']", []), ("not an expected object at [.'$zero']", []), ('unexpected $zero []', [])]
trafo_0.values.json[10]: FAIL [("not an object [.'$zero']", []), ("not an expected object at [.'$zero']", []), ('unexpected $zero []', [])]
trafo_0.values.json[11]: FAIL [("not an object [.'$zero']", []), ("not an expected object at [.'$zero']", []), ('unexpected $zero []', [])]
